### PR TITLE
Add attributes property to RawKernel

### PR DESCRIPTION
**Pull request by grlee77**
It is helpful to be able to query kernel attributes such as the number of registers that a `RawKernel` is using. The implementation in this PR is based on wrapping `cuFuncGetAttributes` from the Driver API.

I think a couple of these attiributes (`maxDynamicSharedSizeBytes` and `preferredShmemCarveout`) did not exist in CUDA 8.0. To deal with this, any attributes that return status `CUDA_ERROR_INVALID_VALUE` get set to -1 to indicate they are unavailable.

**Note:**
I initially tried using `cudaFuncGetAttributes` from the Runtime API (which reads all of the attributes into a structure in one call). However, when I tried passing in the kernel function I always got:
```
CUDARuntimeError: cudaErrorInvalidDeviceFunction: invalid device function
```
Perhaps this is because `extern "C"` is used for the `RawKernel` source, but I am not sure.

### CODE REVIEW

**Comment by grlee77** [comment]
Any feedback regarding this feature? I find it useful to be able to query the number of registers a `RawKernel` uses, for example.

File: cupy/cuda/driver.pyx [hunk 1]
```diff
@@ -227,3 +257,76 @@ cpdef launchKernel(
             shared_mem_bytes, <Stream>stream,
             <void**>kernel_params, <void**>extra)
     check_status(status)
+
+
+cpdef FuncAttributes funcGetAttributes(size_t func):
```
Comment (kmaehashi) [inline cupy/cuda/driver.pyx]: I'd like to keep `driver.pyx` as a simple thin wrapper (1-to-1 mapping) to CUDA Driver API.

File: cupy/core/raw.pyx [hunk 2]
```diff
@@ -51,6 +51,18 @@ cdef class RawKernel:
...
+    def attributes(self):
+        """Returns an object containing runtime kernel attributes.
+
+        Returns:
+            attributes (FuncAttributes): A python class containing the
+                kernel's attributes. For example, ``attributes.numRegs``
+                corresponds to the number of registers used by the kernel.
+        """
+        kern = _get_raw_kernel(self.code, self.name, self.options)
+        return cupy.cuda.driver.funcGetAttributes(kern.ptr)
```
Comment (kmaehashi) [inline cupy/core/raw.pyx]: How about returning attributes as dict, like as done in `cupy.cuda.Device().attributes`?

**Comment by grlee77** [comment]
Thanks @kmaehashi.
I have followed your suggestion and removed the `FuncAttributes` class, replacing it with a `dict` instead. The `dict` is memoized now as well (in `raw.pyx`)

I still need to remove `funcGetAttributes` from driver.pyx if I understand your request correctly, though. I will do that in a separate commit soon.

File: cupy/cuda/driver.pyx (hunk 1 above)
Comment (grlee77) [reply cupy/cuda/driver.pyx]: Okay, I have now replaced this with a simple wrapper `funcGetAttribute` around `cuFuncGetAttribute`.

File: cupy/core/raw.pyx (hunk 2 above)
Comment (grlee77) [reply cupy/core/raw.pyx]: I have updated to construct a dictionary now (within raw.pyx instead of driver.pyx).

**Comment by grlee77** [comment]
Okay, this should now be ready for a new review.

**Comment by grlee77** [comment]
Please review #2120 prior to reviewing this PR again. That PR is an expansion of this one and adds support for setting attributes as well. I think most likely this PR can be closed in favor of that one.

**Comment by grlee77** [comment]
closed in favor of  updated version in #2369

**Pull request closed without merging**